File: src/main/java/org/railwaystations/rsapi/services/ProfileService.java
package org.railwaystations.rsapi.services;

import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.lang3.StringUtils;
import org.railwaystations.rsapi.adapter.db.UserDao;
import org.railwaystations.rsapi.domain.model.User;
import org.railwaystations.rsapi.domain.port.out.Mailer;
import org.railwaystations.rsapi.domain.port.out.Monitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import javax.validation.constraints.NotNull;
import java.util.Optional;
import java.util.UUID;

@Service
public class ProfileService {

    private static final Logger LOG = LoggerFactory.getLogger(ProfileService.class);

    private final Monitor monitor;
    private final Mailer mailer;
    private final UserDao userDao;
    private final String eMailVerificationUrl;
    private final PasswordEncoder passwordEncoder;

    public ProfileService(final Monitor monitor, final Mailer mailer, final UserDao userDao, @Value("${mailVerificationUrl}") final String eMailVerificationUrl, final PasswordEncoder passwordEncoder) {
        this.monitor = monitor;
        this.mailer = mailer;
        this.userDao = userDao;
        this.eMailVerificationUrl = eMailVerificationUrl;
        this.passwordEncoder = passwordEncoder;
    }

    public void changePassword(final User user, final String newPassword) {
        LOG.info("Password change for '{}'", user.getEmail());
        final String trimmedPassword = StringUtils.trimToEmpty(newPassword);
        if (trimmedPassword.length() < 8 ) {
            throw new IllegalArgumentException("Password too short");
        }
        userDao.updateCredentials(user.getId(), passwordEncoder.encode(trimmedPassword));
    }

    public User resetPassword(final String nameOrEmail, final String clientInfo) {
        LOG.info("Password reset requested for '{}'", nameOrEmail);
        final User user = userDao.findByEmail(User.normalizeEmail(nameOrEmail))
                .orElse(userDao.findByNormalizedName(User.normalizeName(nameOrEmail)).orElse(null));

        if (user == null) {
            return null;
        }

        if (StringUtils.isBlank(user.getEmail())) {
            monitor.sendMessage(
                    String.format("Can't reset password for '%s' failed: no email available%nvia %s",
                            nameOrEmail, clientInfo));
            throw new IllegalArgumentException(String.format("Email '%s' is empty", user.getEmail()));
        }

        user.setNewPassword(createNewPassword());
        encryptPassword(user);
        userDao.updateCredentials(user.getId(), user.getKey());
        monitor.sendMessage(String.format("Reset Password for '%s', email='%s'", user.getName(), user.getEmail()));

        sendPasswordMail(user);
        if (!user.isEmailVerified()) {
            // if the email is not yet verified, we can verify it with the next login
            userDao.updateEmailVerification(user.getId(), User.EMAIL_VERIFIED_AT_NEXT_LOGIN);
        }
        return user;
    }

    public void register(final User newUser, final String clientInfo) throws ProfileConflictException {
        LOG.info("New registration for '{}' with '{}'", newUser.getName(), newUser.getEmail());

        if (!newUser.isValidForRegistration()) {
            throw new IllegalArgumentException("Invalid data");
        }

        final Optional<User> existingName = userDao.findByNormalizedName(newUser.getNormalizedName());
        if (existingName.isPresent() && !newUser.getEmail().equals(existingName.get().getEmail())) {
            monitor.sendMessage(
                    String.format("Registration for user '%s' with eMail '%s' failed, name is already taken by different eMail '%s'%nvia %s",
                            newUser.getName(), newUser.getEmail(), existingName.get().getEmail(), clientInfo));
            throw new ProfileConflictException("Name is already taken by different eMail");
        }

        if (userDao.findByEmail(newUser.getEmail()).isPresent()) {
            monitor.sendMessage(
                    String.format("Registration for user '%s' with eMail '%s' failed, eMail is already taken%nvia %s",
                            newUser.getName(), newUser.getEmail(), clientInfo));
            throw new ProfileConflictException("eMail is already taken");
        }

        final boolean passwordProvided = StringUtils.isNotBlank(newUser.getNewPassword());
        if (!passwordProvided) {
            newUser.setNewPassword(createNewPassword());
            newUser.setEmailVerification(User.EMAIL_VERIFIED_AT_NEXT_LOGIN);
        } else {
            newUser.setEmailVerificationToken(UUID.randomUUID().toString());
        }

        encryptPassword(newUser);
        saveRegistration(newUser);

        if (passwordProvided) {
            sendEmailVerification(newUser);
        } else {
            sendPasswordMail(newUser);
        }

        monitor.sendMessage(
                String.format("New registration{nickname='%s', email='%s', license='%s', photoOwner=%s, link='%s', anonymous=%s}%nvia %s",
                        newUser.getName(), newUser.getEmail(), newUser.getLicense(), newUser.isOwnPhotos(),
                        newUser.getUrl(), newUser.isAnonymous(), clientInfo));
    }

    private String createNewPassword() {
        return RandomStringUtils.randomAlphanumeric(12);
    }

    private void encryptPassword(@NotNull final User user) {
        user.setKey(passwordEncoder.encode(user.getNewPassword()));
    }

    public void updateProfile(final User user, final User newProfile, final String clientInfo) throws ProfileConflictException {
        LOG.info("Update profile for '{}'", user.getEmail());

        if (!newProfile.isValid()) {
            LOG.info("Update Profile failed: User invalid {}", newProfile);
            throw new IllegalArgumentException();
        }

        if (!newProfile.getNormalizedName().equals(user.getNormalizedName())) {
            if (userDao.findByNormalizedName(newProfile.getNormalizedName()).isPresent()) {
                LOG.info("Name conflict '{}'", newProfile.getName());
                throw new ProfileConflictException("Name conflict");
            }
            monitor.sendMessage(
                    String.format("Update nickname for user '%s' to '%s'%nvia%s",
                            user.getName(), newProfile.getName(), clientInfo));
        }

        if (!newProfile.getEmail().equals(user.getEmail())) {
            if (userDao.findByEmail(newProfile.getEmail()).isPresent()) {
                LOG.info("Email conflict '{}'", newProfile.getEmail());
                throw new ProfileConflictException("Email conflict");
            }
            newProfile.setEmailVerificationToken(UUID.randomUUID().toString());
            monitor.sendMessage(
                    String.format("Update email for user '%s' from email '%s' to '%s'%nvia%s",
                            user.getName(), user.getEmail(), newProfile.getEmail(), clientInfo));
            sendEmailVerification(newProfile);
        } else {
            // keep email verification status
            newProfile.setEmailVerification(user.getEmailVerification());
        }

        newProfile.setId(user.getId());
        userDao.update(newProfile);
    }

    public void resendEmailVerification(final User user) {
        LOG.info("Resend EmailVerification for '{}'", user.getEmail());
        user.setEmailVerificationToken(UUID.randomUUID().toString());
        userDao.updateEmailVerification(user.getId(), user.getEmailVerification());
        sendEmailVerification(user);
    }

    public Optional<User> emailVerification(final String token) {
        final Optional<User> userByToken = userDao.findByEmailVerification(User.EMAIL_VERIFICATION_TOKEN + token);
        if (userByToken.isPresent()) {
            final User user = userByToken.get();
            userDao.updateEmailVerification(user.getId(), User.EMAIL_VERIFIED);
            monitor.sendMessage(
                    String.format("Email verified {nickname='%s', email='%s'}", user.getName(), user.getEmail()));
        }
        return userByToken;
    }

    private void sendPasswordMail(@NotNull final User user) {
        final String text = String.format("Hello,%n%n" +
                        "your new password is: %1$s%n%n" +
                        "Cheers%n" +
                        "Your Railway-Stations-Team%n" +
                        "%n---%n" +
                        "Hallo,%n%n" +
                        "Dein neues Passwort lautet: %1$s%n%n" +
                        "Viele Grüße%n" +
                        "Dein Bahnhofsfoto-Team", user.getNewPassword());
        mailer.send(user.getEmail(), "Railway-Stations.org new password", text);
        LOG.info("Password sent to {}", user.getEmail());
    }

    private void sendEmailVerification(@NotNull final User user) {
        final String url = eMailVerificationUrl + user.getEmailVerificationToken();
        final String text = String.format("Hello,%n%n" +
                        "please click on %1$s to verify your eMail-Address.%n%n" +
                        "Cheers%n" +
                        "Your Railway-Stations-Team%n" +
                        "%n---%n" +
                        "Hallo,%n%n" +
                        "bitte klicke auf %1$s, um Deine eMail-Adresse zu verifizieren%n%n" +
                        "Viele Grüße%n" +
                        "Dein Bahnhofsfoto-Team", url);
        mailer.send(user.getEmail(), "Railway-Stations.org eMail verification", text);
        LOG.info("Email verification sent to {}", user.getEmail());
    }

    private void saveRegistration(final User registration) {
        final Integer id = userDao.insert(registration);
        LOG.info("User '{}' created with id {}", registration.getName(), id);
    }

    public static class ProfileConflictException extends Exception {

        public ProfileConflictException(final String message) {
            super(message);
        }

    }

}

File: core/src/main/kotlin/org/railwaystations/rsapi/core/services/ProfileService.kt
package org.railwaystations.rsapi.core.services

import org.railwaystations.rsapi.core.model.*
import org.railwaystations.rsapi.core.ports.inbound.ManageProfileUseCase
import org.railwaystations.rsapi.core.ports.inbound.ManageProfileUseCase.ProfileConflictException
import org.railwaystations.rsapi.core.ports.outbound.MailerPort
import org.railwaystations.rsapi.core.ports.outbound.MonitorPort
import org.railwaystations.rsapi.core.ports.outbound.OAuth2AuthorizationPort
import org.railwaystations.rsapi.core.ports.outbound.UserPort
import org.railwaystations.rsapi.core.utils.Logger
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.MessageSource
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.util.*

@Service
class ProfileService(
    private val monitorPort: MonitorPort,
    private val mailerPort: MailerPort,
    private val userPort: UserPort,
    private val authorizationPort: OAuth2AuthorizationPort,
    @param:Value(
        "\${mailVerificationUrl}"
    ) private val eMailVerificationUrl: String,
    private val passwordEncoder: PasswordEncoder,
    private val messageSource: MessageSource
) : ManageProfileUseCase {

    private val log by Logger()

    override fun changePassword(user: User, newPassword: String) {
        log.info("Password change for '{}'", user.email)
        val trimmedPassword = newPassword.trim()
        require(trimmedPassword.length >= 8) { "Password too short" }
        userPort.updateCredentials(user.id, passwordEncoder.encode(trimmedPassword))
        authorizationPort.deleteAllByUser(user.name)
    }

    override fun resetPassword(nameOrEmail: String, clientInfo: String?) {
        log.info("Password reset requested for '{}'", nameOrEmail)
        val user = userPort.findByEmail(normalizeEmail(nameOrEmail))
            ?: userPort.findByName(nameOrEmail)

        requireNotNull(user) { "Can't reset password for unknown user" }

        if (user.email.isNullOrBlank()) {
            monitorPort.sendMessage("Can't reset password for '$nameOrEmail' failed: no email available\nvia $clientInfo")
            throw IllegalArgumentException("Email '${user.email}' is empty")
        }

        val newPassword = createNewPassword()
        val key = encryptPassword(newPassword)
        userPort.updateCredentials(user.id, key)
        monitorPort.sendMessage("Reset Password for '${user.name}', email='${user.email}'")

        sendPasswordMail(user.email, newPassword, user.locale)
        if (!user.isEmailVerified) {
            // if the email is not yet verified, we can verify it with the next login
            userPort.updateEmailVerification(user.id, EMAIL_VERIFIED_AT_NEXT_LOGIN)
        }
        authorizationPort.deleteAllByUser(user.name)
    }

    override fun register(newUser: User, clientInfo: String?) {
        log.info("New registration for '{}' with '{}'", newUser.name, newUser.email)

        require(newUser.isValid) { "Invalid data" }

        val existingName = userPort.findByName(newUser.name)
        existingName?.let {
            if (newUser.email != it.email) {
                monitorPort.sendMessage(
                    "Registration for user '${newUser.name}' with eMail '${newUser.email}' failed, name is already taken by different eMail '${it.email}'\nvia $clientInfo"
                )
                throw ProfileConflictException()
            }
        }

        if (userPort.countBlockedUsername(newUser.name) != 0) {
            monitorPort.sendMessage(
                "Registration for user '${newUser.name}' with eMail '${newUser.email}' failed, name is blocked\nvia $clientInfo"
            )
            throw ProfileConflictException()
        }

        userPort.findByEmail(newUser.email!!)?.let {
            monitorPort.sendMessage(
                "Registration for user '${newUser.name}' with eMail '${newUser.email}' failed, eMail is already taken\nvia $clientInfo"
            )
            throw ProfileConflictException()
        }

        var password = newUser.newPassword
        var emailVerificationToken = createNewEmailVerificationToken()
        val passwordProvided = !password.isNullOrBlank()
        if (!passwordProvided) {
            password = createNewPassword()
            emailVerificationToken = EMAIL_VERIFIED_AT_NEXT_LOGIN
        }

        val key = encryptPassword(password)
        saveRegistration(newUser, key, emailVerificationToken)

        if (passwordProvided) {
            sendEmailVerification(newUser.email, emailVerificationToken)
        } else {
            sendPasswordMail(newUser.email, password, newUser.locale)
        }

        monitorPort.sendMessage(
            """
                New registration{nickname='${newUser.name}', email='${newUser.email}'}
                via $clientInfo
            """.trimIndent()
        )
    }

    private fun encryptPassword(password: String?): String {
        return passwordEncoder.encode(password)
    }

    override fun updateProfile(user: User, newProfile: User, clientInfo: String?) {
        log.info("Update profile for '{}'", user.email)

        if (!newProfile.isValid) {
            log.info("Update Profile failed: User invalid {}", newProfile)
            throw IllegalArgumentException()
        }

        if (newProfile.name != user.name) {
            userPort.findByName(newProfile.name)?.let {
                log.info("Name conflict '{}'", newProfile.name)
                throw ProfileConflictException()
            }
            monitorPort.sendMessage("Update nickname for user '${user.name}' to '${newProfile.name}'\nvia $clientInfo")
        }

        if (newProfile.email != user.email) {
            userPort.findByEmail(newProfile.email!!)?.let {
                log.info("Email conflict '{}'", newProfile.email)
                throw ProfileConflictException()
            }
            monitorPort.sendMessage(
                "Update email for user '${user.name}' from email '${user.email}' to '${newProfile.email}'\nvia $clientInfo"
            )
            val emailVerificationToken = createNewEmailVerificationToken()
            sendEmailVerification(newProfile.email, emailVerificationToken)
            userPort.updateEmailVerification(user.id, emailVerificationToken)
        }

        userPort.update(user.id, newProfile)
    }

    override fun resendEmailVerification(user: User) {
        log.info("Resend EmailVerification for '{}'", user.email)
        val emailVerificationToken = createNewEmailVerificationToken()
        userPort.updateEmailVerification(user.id, emailVerificationToken)
        sendEmailVerification(user.email!!, emailVerificationToken)
    }

    override fun emailVerification(token: String): User? {
        return userPort.findByEmailVerification(token)
            ?.let { user: User ->
                userPort.updateEmailVerification(user.id, EMAIL_VERIFIED)
                monitorPort.sendMessage("Email verified {nickname='${user.name}', email='${user.email}'}")
                user
            }
    }

    override fun deleteProfile(user: User, userAgent: String?) {
        userPort.anonymizeUser(user.id)
        userPort.addUsernameToBlocklist(user.name)
        authorizationPort.deleteAllByUser(user.name)
        monitorPort.sendMessage("Closed account ${user.id} - ${user.name}")
    }

    override fun updateLocale(user: User, locale: Locale) {
        userPort.updateLocale(user.id, locale.toLanguageTag())
    }

    private fun sendPasswordMail(email: String, newPassword: String, locale: Locale) {
        val text = messageSource.getMessage("password_mail", arrayOf(newPassword), locale)
        mailerPort.send(email, "Railway-Stations.org new password", text)
        log.info("Password sent to {}", email)
    }

    private fun sendEmailVerification(email: String, emailVerificationToken: String) {
        val url = eMailVerificationUrl + emailVerificationToken
        val text = """
                Hello,
                                        
                please click on $url to verify your eMail-Address.
                                        
                Cheers
                Your Railway-Stations-Team
                                        
                ---
                Hallo,
                                        
                bitte klicke auf $url, um Deine eMail-Adresse zu verifizieren.
                                        
                Viele Grüße
                Dein Bahnhofsfoto-Team
                """.trimIndent()
        mailerPort.send(email, "Railway-Stations.org eMail verification", text)
        log.info("Email verification sent to {}", email)
    }

    private fun saveRegistration(registration: User, key: String, emailVerification: String) {
        val id = userPort.insert(registration, key, emailVerification)
        log.info("User '{}' created with id {}", registration.name, id)
    }
}

private val possibleCharacters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789~`!@#$%^&*()-_=+[{]}\\|;:\'\",<.>/?".toCharArray()

private fun createNewPassword() = List(20) { possibleCharacters.random() }.joinToString("")
